### PR TITLE
Blog Post on Flink 1.10 Python UDF support

### PR DESCRIPTION
**Pull request by MarkSfik**
Jincheng and I added a new blogpost for adding support for Python UDFs in Flink 1.10.

### CODE REVIEW

**Comment by MarkSfik** [comment]
This is the latest version of the blog post that @sunjincheng121 and I worked on about the python UDF support in Flink 1.10 

cc @fhueske

File: _posts/2020-02-23-pyflink-udf-support-flink.md [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+
+# How to use PyFlink with UDFs in Flink 1.10
+
+This section provides some Python user defined function (UDF) examples, including how to install PyFlink, how to define/register/invoke UDFs in PyFlink and how to execute the job.
+
+
+### Install PyFlink
+Using Python in Apache Flink requires installing PyFlink. PyFlink is available through PyPi and can be easily installed using pip: 
+
+```python
```
Comment (fhueske) [inline _posts/2020-02-23-pyflink-udf-support-flink.md]: The installation commands are missing here (and maybe something else?)

File: _posts/2020-02-23-pyflink-udf-support-flink.md [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+  * The Python user defined function can also aggregate the logging and metrics to the Python UDF Runner via the gRPC logging service and the metrics service during execution.
+  * The execution results are finally sent to the Python UDF Runner via the gRPC data service.
+
+
+# How to use PyFlink with UDFs in Flink 1.10
+
+This section provides some Python user defined function (UDF) examples, including how to install PyFlink, how to define/register/invoke UDFs in PyFlink and how to execute the job.
+
+
+### Install PyFlink
```
Comment (fhueske) [inline _posts/2020-02-23-pyflink-udf-support-flink.md]: Move to heading level 2?

File: _posts/2020-02-23-pyflink-udf-support-flink.md (hunk 1 above)
Comment (fhueske) [reply _posts/2020-02-23-pyflink-udf-support-flink.md]: Add a heading for how to use Python UDFs

Comment (fhueske) [reply _posts/2020-02-23-pyflink-udf-support-flink.md]: Maybe first show a simple but complete Pyflink program without a UDF?

File: _posts/2020-02-23-pyflink-udf-support-flink.md [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+
+The command builds and runs the Python Table API program in a local mini-cluster. You can also submit the Python Table API program to a remote cluster using different command lines, (see more details [here](https://ci.apache.org/projects/flink/flink-docs-master/ops/cli.html#job-submission-examples)).
+
+Finally, you can see the execution result on the command line:
+
+`$ cat /tmp/output
+ 3`
+
+
+### Python UDF dependency management
```
Comment (fhueske) [inline _posts/2020-02-23-pyflink-udf-support-flink.md]: move this to heading 2 for more structure?

File: _posts/2020-02-23-pyflink-udf-support-flink.md [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+    .create_temporary_table('mySink')
+
+t_env.from_path('mySource')\
+    .select("add(a, b)") \
+    .insert_into('mySink')
+
+t_env.execute("tutorial_job")
+```
+
+### Submit the job
```
Comment (fhueske) [inline _posts/2020-02-23-pyflink-udf-support-flink.md]: we had the same heading before. This is a bit confusing

File: _posts/2020-02-23-pyflink-udf-support-flink.md [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+date: 2020-02-23T12:00:00.000Z
+authors:
+- Jincheng:
+  name: "Jincheng Sun"
+excerpt: Flink1.10 extends its support for Python by adding Python UDFs in PyFlink. This post explains how UDFs work in PyFlink and gives some practical examples of how to use UDFs in PyFlink. 
+---
+
+Flink 1.9 introduced the Python Table API, allowing developers and data engineers to write Python Table API jobs for Table Transformations and Analysis, such as Python ETL or Aggregate jobs. However, Python users faced some limitations when it came to support for Python UDFs in Flink 1.9, preventing them from extending the system’s built-in functionality. 
+
+In Flink1.10, the community further extended the support for Python by adding Python UDFs in PyFlink. Additionally, both the Python UDF Environment and Dependency Management are now supported, allowing users to import third-party libraries in the UDFs, leveraging Python's rich set of third-party libraries.
```
Comment (sjwiesman) [inline _posts/2020-02-23-pyflink-udf-support-flink.md]:
```suggestion
In Flink 1.10, the community further extended the support for Python by adding Python UDFs in PyFlink. Additionally, both the Python UDF Environment and Dependency Management are now supported, allowing users to import third-party libraries in the UDFs, leveraging Python's rich set of third-party libraries. 
```

File: _posts/2020-02-23-pyflink-udf-support-flink.md (hunk 1 above)
Comment (sunjincheng121) [reply _posts/2020-02-23-pyflink-udf-support-flink.md]: > The installation commands are missing here (and maybe something else?)

+1, I think the command `python -m pip install apache-flink`  should be added  .  same as `Install PyFlink` part in [1].
[1] https://docs.google.com/document/d/1I83o8e4CSu1baKPK9SKq1S1GN-p98AVvDOIgUHQVOZo/edit# 

> Add a heading for how to use Python UDFs

This post is talk about Python UDF support, So I am fine without a simple example. :)

File: _posts/2020-02-23-pyflink-udf-support-flink.md (hunk 4 above)
Comment (sunjincheng121) [reply _posts/2020-02-23-pyflink-udf-support-flink.md]: May be `Submit a job with dependency` ?

File: _posts/2020-02-23-pyflink-udf-support-flink.md (hunk 2 above)
Comment (MarkSfik) [inline _posts/2020-02-23-pyflink-udf-support-flink.md]:
```suggestion
## Install PyFlink
```

File: _posts/2020-02-23-pyflink-udf-support-flink.md (hunk 3 above)
Comment (MarkSfik) [inline _posts/2020-02-23-pyflink-udf-support-flink.md]:
```suggestion
## Python UDF dependency management
```

File: _posts/2020-02-23-pyflink-udf-support-flink.md (hunk 4 above)
Comment (MarkSfik) [inline _posts/2020-02-23-pyflink-udf-support-flink.md]:
```suggestion
### Submit a job with dependency 
```

**Comment by sunjincheng121** [comment]
Thanks @MarkSfik and @fhueske , I would like to merging this PR after add a few minor improvement changes.

File: _posts/2020-02-23-pyflink-udf-support-flink.md [hunk 6]
```diff
@@ -0,0 +1,280 @@
+---
+layout: post
+title: "PyFlink: Introducing Python Support for UDFs in Flink's Table API"
+date: 2020-02-23T12:00:00.000Z
+authors:
+- Jincheng:
+  name: "Jincheng Sun"
+excerpt: Flink1.10 extends its support for Python by adding Python UDFs in PyFlink. This post explains how UDFs work in PyFlink and gives some practical examples of how to use UDFs in PyFlink.
```
Comment (rmetzger) [inline _posts/2020-02-23-pyflink-udf-support-flink.md]:
```suggestion
excerpt: Flink 1.10 extends its support for Python by adding Python UDFs in PyFlink. This post explains how UDFs work in PyFlink and gives some practical examples of how to use UDFs in PyFlink. 
```

File: _posts/2020-02-23-pyflink-udf-support-flink.md [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+layout: post
+title: "PyFlink: Introducing Python Support for UDFs in Flink's Table API"
+date: 2020-02-23T12:00:00.000Z
+authors:
+- Jincheng:
+  name: "Jincheng Sun"
+excerpt: Flink1.10 extends its support for Python by adding Python UDFs in PyFlink. This post explains how UDFs work in PyFlink and gives some practical examples of how to use UDFs in PyFlink. 
+---
+
+Flink 1.9 introduced the Python Table API, allowing developers and data engineers to write Python Table API jobs for Table Transformations and Analysis, such as Python ETL or Aggregate jobs. However, Python users faced some limitations when it came to support for Python UDFs in Flink 1.9, preventing them from extending the system’s built-in functionality.
```
Comment (rmetzger) [inline _posts/2020-02-23-pyflink-udf-support-flink.md]:
```suggestion
Flink 1.9 introduced the Python Table API, allowing developers and data engineers to write Python Table API jobs for Table transformations and analysis, such as Python ETL or aggregate jobs. However, Python users faced some limitations when it came to support for Python UDFs in Flink 1.9, preventing them from extending the system’s built-in functionality. 
```

File: _posts/2020-02-23-pyflink-udf-support-flink.md (hunk 5 above)
Comment (rmetzger) [inline _posts/2020-02-23-pyflink-udf-support-flink.md]:
```suggestion
In Flink 1.10, the community further extended the support for Python by adding Python UDFs in PyFlink. Additionally, both the Python UDF environment and dependency management are now supported, allowing users to import third-party libraries in the UDFs, leveraging Python's rich set of third-party libraries. 
```

**Comment by MarkSfik** [comment]
Thanks for the comments and suggestions @rmetzger 
I merged the updates to this branch. 

Let me know if this is good to be merged? 

cc @sunjincheng121  cc @fhueske

**Comment by sunjincheng121** [comment]
I have had some updates on Python UDF recently. I plan to make some improvements by referring to the Chinese blog[1] I wrote a few days ago. @MarkSfik 

[1] https://enjoyment.cool/2020/03/16/Apache-Flink-%E8%A7%86%E9%A2%91%E7%B3%BB%E5%88%97-%E5%BF%85%E4%BF%AE%E8%AF%BE%EF%BC%81%E4%B8%80%E5%B0%8F%E6%97%B6%E5%90%83%E9%80%8FPyFlink/

**Comment by MarkSfik** [comment]
Thanks @sunjincheng121 
Let me know once you have made the changes so I can review and adjust the post accordingly! 
Thanks

**Comment by sunjincheng121** [comment]
Thanks @MarkSfik , I do not know how to add the commit to your branch. So, I have opened a new PR here: https://github.com/apache/flink-web/pull/316

**Pull request closed without merging**